File: apps/server/src/recipes.ts
import type { Recipe, MaterialReq, QualityBand } from "@divineruin/shared";
import { sql } from "./db.ts";

// DB-loaded recipe content (M5.1). Mirrors the training-programs loader in
// activity_templates.ts: content/recipes.json -> recipes table, loaded at
// startup, parsed fail-loud, exposed via getRecipe/listRecipes. Both the TS
// server and the Python agent read recipes from the DB (constraint 8508fdb1abc3,
// wisdom bb73edd9b94d) — never a hardcoded TS const.

const RECIPE_CATEGORIES = new Set<Recipe["category"]>([
  "weapon",
  "armor",
  "consumable",
  "tool",
  "enchantment",
  "ammunition",
]);
const RECIPE_TIERS = new Set<Recipe["tier"]>(["basic", "trained", "expert", "master"]);
const WORKSPACES = new Set<Recipe["workspace_required"]>([
  "field",
  "workshop",
  "forge",
  "laboratory",
]);
const MATERIAL_TIERS = new Set<MaterialReq["tier_minimum"]>([1, 2, 3, 4]);
// Canonical narration quality bands (crafting-narration-bands). The loader fails
// loud on any other band key so a content typo can't silently miss at runtime.
const QUALITY_BANDS = new Set<QualityBand>(["exceptional", "success", "partial", "failure"]);

/**
 * Validate a count field is a real integer in [min, ∞). Count fields feed
 * Array(n).fill (recipeMaterialIds) and arithmetic; a float, negative, NaN, or
 * Infinity row would throw RangeError or silently corrupt downstream — so the
 * fail-loud parser rejects them at the boundary rather than at craft time.
 */
function requireIntAtLeast(value: unknown, min: number, ctx: string): number {
  if (typeof value !== "number" || !Number.isInteger(value) || value < min) {
    throw new Error(`${ctx} ${String(value)} is not an integer >= ${min}`);
  }
  return value;
}

// Runtime-loaded recipes (populated by loadRecipes at startup).
let recipes: ReadonlyMap<string, Recipe> = new Map();

export function getRecipe(id: string): Recipe | undefined {
  return recipes.get(id);
}

export function listRecipes(): Recipe[] {
  return Array.from(recipes.values());
}

export function setRecipes(map: ReadonlyMap<string, Recipe>): void {
  recipes = map;
}

// Real-world async wait derived from async_cycles (the spec's crafting-duration
// field): 4h per cycle, with a 15-min floor for instant (0-cycle) field recipes.
// max = 2*min preserves the ~2x spread of the old hand-tuned durations.
const CRAFT_CYCLE_SECONDS = 14400; // 4h
const CRAFT_FLOOR_SECONDS = 900; // 15min

/** Flatten a recipe's MaterialReq[] into a material-id list, repeated by quantity. */
export function recipeMaterialIds(recipe: Recipe): string[] {
  return recipe.materials.flatMap((m) => Array<string>(m.quantity).fill(m.material_id));
}

/** Crafting-activity real-world wait window (seconds) from the recipe's async_cycles. */
export function craftingDurationSeconds(recipe: Recipe): { min: number; max: number } {
  const min =
    recipe.async_cycles > 0 ? recipe.async_cycles * CRAFT_CYCLE_SECONDS : CRAFT_FLOOR_SECONDS;
  return { min, max: min * 2 };
}

function parseMaterialReq(raw: unknown, ctx: string): MaterialReq {
  if (!raw || typeof raw !== "object") throw new Error(`${ctx} is not an object`);
  const m = raw as Record<string, unknown>;
  if (typeof m.material_id !== "string" || m.material_id.length === 0) {
    throw new Error(`${ctx}.material_id is not a non-empty string`);
  }
  const quantity = requireIntAtLeast(m.quantity, 1, `${ctx}.quantity`);
  if (typeof m.tier_minimum !== "number" || !MATERIAL_TIERS.has(m.tier_minimum as 1 | 2 | 3 | 4)) {
    throw new Error(`${ctx}.tier_minimum ${String(m.tier_minimum)} is not 1-4`);
  }
  if (typeof m.substitutable !== "boolean") {
    throw new Error(`${ctx}.substitutable is not a boolean`);
  }
  return {
    material_id: m.material_id,
    quantity,
    tier_minimum: m.tier_minimum as MaterialReq["tier_minimum"],
    substitutable: m.substitutable,
  };
}

/**
 * Canonical fail-loud parser for one recipes-table row. Validates all 16 Recipe
 * fields (every one required — recipes are fully-specified content, decision
 * cb1d48971cc2); throws with a `recipes[<id>].<field>` context on any mismatch.
 * Validates a single recipe in isolation — material_id->catalog referential
 * integrity is a cross-entity check owned by the content-validation test.
 */
export function parseRecipeRow(id: string, raw: unknown): Recipe {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error(`recipes[${id}].data is not an object`);
  }
  const data = raw as Record<string, unknown>;
  const ctx = `recipes[${id}]`;

  if (typeof data.name !== "string") throw new Error(`${ctx}.name is not a string`);
  if (
    typeof data.category !== "string" ||
    !RECIPE_CATEGORIES.has(data.category as Recipe["category"])
  ) {
    throw new Error(`${ctx}.category ${String(data.category)} is invalid`);
  }
  if (typeof data.tier !== "string" || !RECIPE_TIERS.has(data.tier as Recipe["tier"])) {
    throw new Error(`${ctx}.tier ${String(data.tier)} is invalid`);
  }
  if (!Array.isArray(data.materials)) throw new Error(`${ctx}.materials is not an array`);
  const materials = data.materials.map((m, i) => parseMaterialReq(m, `${ctx}.materials[${i}]`));
  if (!Array.isArray(data.optional_materials)) {
    throw new Error(`${ctx}.optional_materials is not an array`);
  }
  const optionalMaterials = data.optional_materials.map((m, i) =>
    parseMaterialReq(m, `${ctx}.optional_materials[${i}]`),
  );
  if (typeof data.tainted_materials !== "boolean") {
    throw new Error(`${ctx}.tainted_materials is not a boolean`);
  }
  if (
    typeof data.workspace_required !== "string" ||
    !WORKSPACES.has(data.workspace_required as Recipe["workspace_required"])
  ) {
    throw new Error(`${ctx}.workspace_required ${String(data.workspace_required)} is invalid`);
  }
  const craftingDc = requireIntAtLeast(data.crafting_dc, 1, `${ctx}.crafting_dc`);
  if (typeof data.time !== "string") throw new Error(`${ctx}.time is not a string`);
  const asyncCycles = requireIntAtLeast(data.async_cycles, 0, `${ctx}.async_cycles`);
  if (typeof data.output_item !== "string" || data.output_item.length === 0) {
    throw new Error(`${ctx}.output_item is not a non-empty string`);
  }
  const outputQuantity = requireIntAtLeast(data.output_quantity, 1, `${ctx}.output_quantity`);
  const studyCost = requireIntAtLeast(data.study_cost, 0, `${ctx}.study_cost`);
  if (!Array.isArray(data.discovery_sources)) {
    throw new Error(`${ctx}.discovery_sources is not an array`);
  }
  const discoverySources = data.discovery_sources.map((s, i) => {
    if (typeof s !== "string") throw new Error(`${ctx}.discovery_sources[${i}] is not a string`);
    return s;
  });
  if (!data.narration_cues || typeof data.narration_cues !== "object") {
    throw new Error(`${ctx}.narration_cues is not an object`);
  }
  const cuesRaw = data.narration_cues as Record<string, unknown>;
  if (Object.keys(cuesRaw).length === 0) throw new Error(`${ctx}.narration_cues is empty`);
  const narrationCues: Partial<Record<QualityBand, string>> = {};
  for (const [band, cue] of Object.entries(cuesRaw)) {
    if (!QUALITY_BANDS.has(band as QualityBand)) {
      throw new Error(`${ctx}.narration_cues[${band}] is not a canonical quality band`);
    }
    if (typeof cue !== "string") throw new Error(`${ctx}.narration_cues[${band}] is not a string`);
    narrationCues[band as QualityBand] = cue;
  }
  // crafting-narration-bands: a recipe carries {success, failure} (2) or all 4.
  // success+failure are always required so a resolved success/failure outcome
  // never finds undefined; exceptional+partial are all-or-nothing. Enforced here
  // at the per-turn DB boundary (not just the content test) so a non-seed row
  // (migration, manual fix, future tool) can't silently miss at narration time.
  if (!("success" in narrationCues) || !("failure" in narrationCues)) {
    throw new Error(`${ctx}.narration_cues must include both success and failure bands`);
  }
  const bandCount = Object.keys(narrationCues).length;
  if (bandCount !== 2 && bandCount !== 4) {
    throw new Error(
      `${ctx}.narration_cues has ${bandCount} bands; expected 2 (success/failure) or all 4`,
    );
  }

  return {
    id,
    name: data.name,
    category: data.category as Recipe["category"],
    tier: data.tier as Recipe["tier"],
    materials,
    optional_materials: optionalMaterials,
    tainted_materials: data.tainted_materials,
    workspace_required: data.workspace_required as Recipe["workspace_required"],
    crafting_dc: craftingDc,
    time: data.time,
    async_cycles: asyncCycles,
    output_item: data.output_item,
    output_quantity: outputQuantity,
    study_cost: studyCost,
    discovery_sources: discoverySources,
    narration_cues: narrationCues,
  };
}

export async function loadRecipes(): Promise<void> {
  const rows = await sql<{ id: string; data: unknown }[]>`
    SELECT id, data FROM recipes
  `;
  const map = new Map<string, Recipe>();
  for (const row of rows) {
    map.set(row.id, parseRecipeRow(row.id, row.data));
  }
  recipes = map;
  console.log(`Loaded ${map.size} recipes`);
}

File: apps/agent/recipes.py
"""Cached recipe accessors for the DM agent (M5.1 crafting).

The Python agent reads recipes from the DB (constraint 8508fdb1abc3, wisdom
bb73edd9b94d) — never a hardcoded const. This mirrors the TS server's
apps/server/src/recipes.ts: a fail-loud parse_recipe_row validating all 16
Recipe fields (every one required — recipes are fully-specified content), plus
cached get_recipe / list_recipes accessors keyed recipe:<id> and recipes:all.

Read-only from the agent's view: only the TS server writes recipes during
resolution. Consumed by story-006's recipe tools (query_recipe_requirements,
learn_recipe), which read the snake_case parsed dict returned here.
"""

import json
import logging

import db

logger = logging.getLogger("divineruin.recipes")

_CATEGORIES = {"weapon", "armor", "consumable", "tool", "enchantment", "ammunition"}
_TIERS = {"basic", "trained", "expert", "master"}
_WORKSPACES = {"field", "workshop", "forge", "laboratory"}
_MATERIAL_TIERS = {1, 2, 3, 4}


def _require_int_at_least(value: object, min_: int, ctx: str) -> int:
    """Mirror TS requireIntAtLeast: a real int >= min_. bool is an int subclass
    in Python, so reject it explicitly — a count field is never True/False."""
    if isinstance(value, bool) or not isinstance(value, int) or value < min_:
        raise ValueError(f"{ctx} {value!r} is not an integer >= {min_}")
    return value


def _parse_material_req(raw: object, ctx: str) -> dict:
    """Mirror TS parseMaterialReq: validate one MaterialReq in isolation."""
    if not isinstance(raw, dict):
        raise ValueError(f"{ctx} is not an object")
    material_id = raw.get("material_id")
    if not isinstance(material_id, str) or len(material_id) == 0:
        raise ValueError(f"{ctx}.material_id is not a non-empty string")
    quantity = _require_int_at_least(raw.get("quantity"), 1, f"{ctx}.quantity")
    tier_minimum = raw.get("tier_minimum")
    if isinstance(tier_minimum, bool) or not isinstance(tier_minimum, int) or tier_minimum not in _MATERIAL_TIERS:
        raise ValueError(f"{ctx}.tier_minimum {tier_minimum!r} is not 1-4")
    substitutable = raw.get("substitutable")
    if not isinstance(substitutable, bool):
        raise ValueError(f"{ctx}.substitutable is not a boolean")
    return {
        "material_id": material_id,
        "quantity": quantity,
        "tier_minimum": tier_minimum,
        "substitutable": substitutable,
    }


def parse_recipe_row(recipe_id: str, raw: object) -> dict:
    """Canonical fail-loud parser for one recipes-table row's `data` payload.

    Mirrors apps/server/src/recipes.ts parseRecipeRow: validates all 16 Recipe
    fields, raising ValueError with a `recipes[<id>].<field>` context on any
    mismatch. id comes from the row key (passed in), not the data. Validates a
    single recipe in isolation — material_id->catalog referential integrity is a
    cross-entity check owned by the TS content-validation test.

    Like the TS parser, narration_cues band keys are NOT enum-validated here
    (concern 31c6bd30ca97, deferred for both languages); only that the dict is
    non-empty with string values. The crafting-narration-bands decision records
    the canonical band vocabulary for future drift-catching.
    """
    ctx = f"recipes[{recipe_id}]"
    if not isinstance(raw, dict):
        raise ValueError(f"{ctx}.data is not an object")

    name = raw.get("name")
    if not isinstance(name, str):
        raise ValueError(f"{ctx}.name is not a string")
    category = raw.get("category")
    if not isinstance(category, str) or category not in _CATEGORIES:
        raise ValueError(f"{ctx}.category {category!r} is invalid")
    tier = raw.get("tier")
    if not isinstance(tier, str) or tier not in _TIERS:
        raise ValueError(f"{ctx}.tier {tier!r} is invalid")

    materials_raw = raw.get("materials")
    if not isinstance(materials_raw, list):
        raise ValueError(f"{ctx}.materials is not an array")
    materials = [_parse_material_req(m, f"{ctx}.materials[{i}]") for i, m in enumerate(materials_raw)]
    optional_raw = raw.get("optional_materials")
    if not isinstance(optional_raw, list):
        raise ValueError(f"{ctx}.optional_materials is not an array")
    optional_materials = [_parse_material_req(m, f"{ctx}.optional_materials[{i}]") for i, m in enumerate(optional_raw)]

    tainted_materials = raw.get("tainted_materials")
    if not isinstance(tainted_materials, bool):
        raise ValueError(f"{ctx}.tainted_materials is not a boolean")
    workspace = raw.get("workspace_required")
    if not isinstance(workspace, str) or workspace not in _WORKSPACES:
        raise ValueError(f"{ctx}.workspace_required {workspace!r} is invalid")

    crafting_dc = _require_int_at_least(raw.get("crafting_dc"), 1, f"{ctx}.crafting_dc")
    time = raw.get("time")
    if not isinstance(time, str):
        raise ValueError(f"{ctx}.time is not a string")
    async_cycles = _require_int_at_least(raw.get("async_cycles"), 0, f"{ctx}.async_cycles")
    output_item = raw.get("output_item")
    if not isinstance(output_item, str) or len(output_item) == 0:
        raise ValueError(f"{ctx}.output_item is not a non-empty string")
    output_quantity = _require_int_at_least(raw.get("output_quantity"), 1, f"{ctx}.output_quantity")
    study_cost = _require_int_at_least(raw.get("study_cost"), 0, f"{ctx}.study_cost")

    sources_raw = raw.get("discovery_sources")
    if not isinstance(sources_raw, list):
        raise ValueError(f"{ctx}.discovery_sources is not an array")
    discovery_sources = []
    for i, s in enumerate(sources_raw):
        if not isinstance(s, str):
            raise ValueError(f"{ctx}.discovery_sources[{i}] is not a string")
        discovery_sources.append(s)

    cues_raw = raw.get("narration_cues")
    if not isinstance(cues_raw, dict) or len(cues_raw) == 0:
        raise ValueError(f"{ctx}.narration_cues is not a non-empty object")
    narration_cues = {}
    for band, cue in cues_raw.items():
        if not isinstance(cue, str):
            raise ValueError(f"{ctx}.narration_cues[{band}] is not a string")
        narration_cues[band] = cue

    return {
        "id": recipe_id,
        "name": name,
        "category": category,
        "tier": tier,
        "materials": materials,
        "optional_materials": optional_materials,
        "tainted_materials": tainted_materials,
        "workspace_required": workspace,
        "crafting_dc": crafting_dc,
        "time": time,
        "async_cycles": async_cycles,
        "output_item": output_item,
        "output_quantity": output_quantity,
        "study_cost": study_cost,
        "discovery_sources": discovery_sources,
        "narration_cues": narration_cues,
    }


async def get_recipe(recipe_id: str) -> dict | None:
    """Return the parsed recipe for recipe_id, or None if unknown.

    Caches the PARSED dict (parse-once), an intentional departure from the raw
    json caching in db_content_queries — recipes are validated at the boundary,
    so the cache holds the already-validated shape. Round-trips cleanly through
    json since the parser emits only plain dict/list/str/int/bool.
    """
    cache_key = f"recipe:{recipe_id}"
    cached = await db._cache_get(cache_key)
    if cached is not None:
        return json.loads(cached)

    pool = await db.get_pool()
    row = await pool.fetchrow("SELECT data FROM recipes WHERE id = $1", recipe_id)
    if row is None:
        return None

    recipe = parse_recipe_row(recipe_id, json.loads(row["data"]))
    await db._cache_set(cache_key, json.dumps(recipe))
    return recipe


async def list_recipes() -> list[dict]:
    """Return all parsed recipes, cached under recipes:all."""
    cache_key = "recipes:all"
    cached = await db._cache_get(cache_key)
    if cached is not None:
        return json.loads(cached)

    pool = await db.get_pool()
    rows = await pool.fetch("SELECT id, data FROM recipes ORDER BY id")
    parsed = [parse_recipe_row(r["id"], json.loads(r["data"])) for r in rows]
    await db._cache_set(cache_key, json.dumps(parsed))
    return parsed
